Compute m_[uE,vN] & prof_[U,V] in profiles independent of pkg/ecco

antnguyen13 opened the pull request:
## What changes does this PR introduce?
feature: allows pkg/profiles to be compiled and run and produced mooring prof_[U,V] outputs without needing pkg/ecco.


## What is the current behaviour? 
At the moment, pkg/profiles works independently only for prof_[T,S] (and may be other tracers (?)).  When having mooring arrays with prof_[U,V], no model equivalent outputs are calculated for these two quantities unless pkg/ecco is compiled. 


## What is the new behaviour 
In this modification, I put ifndef ALLOW_ECCO block inside 3 files within pkg/profiles, such that if pkg/ecco is not compiled, then global variables m_[uE,vN] are still defined, calculated, and prof_[U,V] can be filled in within pkg/profiles.  

When pkg/ecco is defined, then the old (current) behavior takes place as before, with m_[uE,vN] being defined and calculated within pkg/ecco. 

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)
I do not anticipate any breaking changes.  I suspect I might be the first user to try to use pkg/profiles without pkg/ecco and thus realizing I did not get prof_[U,V] unless I compile with pkg/ecco.

## Other information:
While making this mod, I have also noticed that prof_etan also requires pkg/ecco for computation.  Potentally this could also be changed such that it can pkg/profiles is completely independent of pkg/ecco.

christophernhill commented:
@antnguyen13 thanks for this, couple of questions

  1. why is ```#ifndef ALLOW_ECCO``` needed. Thats not clear to me?

  2. ```profiles_inloop.F_new``` looks like a file that got pushed by mistake?

christophernhill commented:
@gaelforget could you take a look at this - thanks. Chris

antnguyen13 commented:
1. prof_[U,V] in pkg/profiles need m_[uE,vN] .  m_[uE,vN] (the eastward and northward components of the model uvel and vvel are defined only in pkg/ecco/ecco.h and calculated in pkg/ecco/ecco_phys.F .  So currently to access these variables to fill out something in pkg/profiles we would also need to compile pkg/ecco .  Ideally I would have preferred that these two global variables m_[uE,vN] are defined outside of any package so we can just access them.  However, I searched the code and the only packages that need them are pkg/ecco and pkg/profiles.  So, to summarize:
a) if compile pkg/profiles & pkg/ecco, m_[uE,vN] are available
b) if compile only pkg/ecco , m_[uE,vN] are available
c) if compile only pkg/profiles, m_[uE,vN] are NOT available.  
I'd like to fix (c) , and avoid double define and double calculating in case of (a) , hence the #ifndef ALLOW_ECCO , i.e., only in the case of (c) that we define and calculate the variables within pkg/profiles.

2. Ah, must be a mistake, sorry, this is my first time ever checking in something on git!  And I do not see a copy of profiles_inloop.F_new on my local machine, so I'm a bit confused.. Could you suggest how to remove that file if locally you don't see it? I had thought I finished all my edits and removed all extra files before doing git add . and git commit .

jklymak commented:
try `git rm  profiles_inloop.F_new`

antnguyen13 commented:
@jklymak Thanks, but I don't have the file on my local machine, that's why I'm confused.. 
[atnguyen@ profiles]$ git rm profiles_inloop.F_new
fatal: pathspec 'pkg/profiles/profiles_inloop.F_new' did not match any files

jklymak commented:
Backup your local branch, and then in the branch do `git reset --hard origin/separate_profiles_ecco`.  That'll get your branch to be exactly whats on GitHub.  Then `git rm` and `git commit` and `git push`.  There is probably also a way to remove the file directly using git rm, but it seems you've somehow changed locally so maybe safer to start with whats on GitHub as described above.

antnguyen13 commented:
@jklymak Thanks! I got through almost everything, except during the commit/push, when it first i updated the wrong branch, then accidentally created a new branch (very new to git and still very confused about how to point to the right branch when pushing...)

@christophernhill Could you please check if the extra file profiles_inloop.F_new is removed?  These are the steps I used:

[atnguyen@ pkg]$ git remote add upstream https://github.com/MITgcm/MITgcm.git
fatal: remote upstream already exists.
[atnguyen@nansen pkg]$ git fetch upstream
[atnguyen@nansen pkg]$ git reset --hard origin/separate_profiles_ecco
HEAD is now at da0e076 Compute m_[uE,vN] & prof_[U,V] in profiles independent of pkg/ecco
[atnguyen@nansen profiles]$ git rm profiles_inloop.F_new
rm 'pkg/profiles/profiles_inloop.F_new'
[atnguyen@nansen profiles]$ git status
# On branch salt_plume_bugfix
# Changes to be committed:
#   (use "git reset HEAD <file>..." to unstage)
#
#	deleted:    profiles_inloop.F_new
#
[atnguyen@nansen profiles]$ git commit -m "remove not needed file"
[salt_plume_bugfix 50fc386] remove not needed file
 1 file changed, 131 deletions(-)
 delete mode 100644 pkg/profiles/profiles_inloop.F_new
[atnguyen@nansen profiles]$ git push -u origin separate_profiles_ecco
Branch separate_profiles_ecco set up to track remote branch separate_profiles_ecco from origin.
Everything up-to-date
[atnguyen@nansen profiles]$ git status
# On branch salt_plume_bugfix
nothing to commit, working directory clean
[atnguyen@nansen profiles]$ git push -u origin origin/separate_profiles_ecco
Total 0 (delta 0), reused 0 (delta 0)
To https://github.com/antnguyen13/MITgcm.git
 * [new branch]      origin/separate_profiles_ecco -> origin/separate_profiles_ecco
[atnguyen@nansen profiles]$ git checkout separate_profiles_ecco
Switched to branch 'separate_profiles_ecco'
[atnguyen@nansen profiles]$ git push
warning: push.default is unset; its implicit value is changing in
Git 2.0 from 'matching' to 'simple'. To squelch this message
and maintain the current behavior after the default changes, use:
Everything up-to-date
[atnguyen@nansen profiles]$ git status
# On branch separate_profiles_ecco
nothing to commit, working directory clean

(now I think i have an extra branch origin/separate_profile_ecco ? )

jklymak commented:
Just put tripple quotes around the whole thing!

Yeah, not sure what you did there, but I just do `git push origin branchname`.

antnguyen13 commented:
@jklymak  Could you check the commands I typed above again (now with edited tripple backquotes)? I did do ```git commit``` after removing.  I think I was just confused which branch I was on and whether I was pushing to the right branch (or that it created new branches).  I think I now have 2 branches: separate_profiles_ecco and origin/separate_profiles_ecco . Let me check via the browser now.

christophernhill commented:
@antnguyen13 

Re: 1. prof_[U,V] in pkg/profiles need m_[uE,vN] . m_[uE,vN] (the eastward and northward components of the model uvel and vvel) are defined only in pkg/ecco/ecco.h and calculated in pkg/ecco/ecco_phys.F  etc....

I'll get together with Gael on this. If possible, it would be nice not to have it in two different places depending on ```ifndef``` double negative, which can be confusing. Not sure if it is possible though!

edoddridge commented:
@antnguyen13 the `origin/` at the start of the branch name means that it is the version of that branch in your GitHub repository (i.e. in your fork of MITgcm, see @jrscott's lovely figure 5.1 [here](http://mitgcm.readthedocs.io/en/latest/contributing/contributing.html#detailed-guide-for-those-less-familiar-with-git-and-github)). It's not a duplicate.

I'm not sure why you're having such trouble removing the file, The instructions from @jklymak should have worked.

gaelforget commented:
My preference, instead of duplicating code from pkg/ecco to pkg/profiles, would be that you add a test in pkg/profiles that issues a warning (a stop would probably be excessive) when variables defined in pkg/ecco are needed in pkg/profiles but pkg/ecco has not been compiled. The test conditions should be easy to formulate once data.profiles and the *.nc inputs have been scanned (i.e., after profiles_init_fixed; ideally in a new function called profiles_check). User can then decide to compile pkg/ecco or otherwise. BTW I dont think that the memory footprint of compiling pkg/ecco on top of pkg/profiles is significant. 

On a related note, the only gridded fields that are currently being defined in pkg/profiles are prof_etan_mean, prof_theta_mean, and prof_salt_mean -- feel free to remove these while you are it. These are actually never used and only prof_etan_mean, which was once used, even gets assigned ... in pkg/ecco by copying the old "psmean" variable . Additional comments about that: using m_eta rather than psmean, if needed, would be better and can be done without prof_etan_mean; replicating pkg/ecco's time-averaging capabilities in pkg/profiles would not be a great idea either in my opinion.

edoddridge commented:
Hang on, I think I've worked it out - you were on the `salt_plume_bugfix` branch when you did the reset and removed the file. You need to be on the `separate_profiles_ecco` branch. Before you execute the commands Jody suggested, execute `git checkout separate_profiles_ecco`

antnguyen13 commented:
@gaelforget @christophernhill 
I think it would be nice to not have pkg/profiles needing pkg/ecco simply for 2 lines of code that access something readily available in the model (and not something specific to pkg/ecco):  the model's U and V (rotated to East and North, using commonly used subroutine ROTATE_UV2EN_RL) .  

Would it be possible to just name the variable different from m_[uE,vN] and have pkg/profiles always just do this calculation without looking to pkg/ecco?  That is to remove my modification of ```ifndef ALLOW_ECCO```, and simply fill in a new set of variables = rotated velocity inside pkg/profiles?

As you mentioned, it's annoying to have "duplicating" code .  But because this is such a basic calculation that I think regardless how m_[uE,vN] might be modified in ecco_phys.F , for pkg/profiles it's always the same: simply the rotated velocities.

antnguyen13 commented:
@edoddridge 
I think i finally figured that part out after I pushed and it said "now on```salt_plume_bugfix``` branch.  I did re-checkout ```separate_profiles_ecco``` and did re-push the new code in.  I hope everything is in order now and that the extra file is removed?

jklymak commented:
> I hope everything is in order now and that the extra file is removed?

Look under the "Files Changed" tab above.

antnguyen13 commented:
@jklymak oh no, i still see it, ack! ok, i'll sort this out within today. Thanks. I wasn't sure what you meant by "Files Changed" yesterday but now see it in my browser.

jklymak commented:
@antnguyen13 I find this page useful, and mostly follow it to understand `git`:  https://matthew-brett.github.io/pydagogue/gitwash/development_workflow.html#some-other-things-you-might-want-to-do

christophernhill commented:
@antnguyen13 

Re: I think it would be nice to not have pkg/profiles needing pkg/ecco simply for 2 lines of code.... 

Agreed. Not sure what is good way to do this. Jeff has offered to help figure that out with you and Gael. It would be nice to avoid having a gradual accretion of duplicate stuff because ecco, autodiff, prof etc.. are treading on each others toes -- maybe that is not possible, but it would be a good goal.

gaelforget commented:
@antnguyen13: since this thread has grown so extensive already, I feel it may be worth asking: why is not compiling pkg/ecco/ important to you? Would this save you a large memory footprint even if data.ecco were empty? Something else?

@christophernhill, regarding "would be nice to avoid having a gradual accretion of duplicate stuff": yes that is my basic concern; replicating the m_eta computation and array in pkg/profiles/ would likely be next; and so on and so forth.

@christophernhill, regarding "ecco, autodiff, prof etc.. are treading on each others toes": I feel that ecco dealing with gridded data sets and profiles dealing with ungridded data sets makes sense as a general principle. The addition of ecco_phys.F was meant to provide a location for quantities like m_eta, m_UE, m_VN to be computed and stored within ecco/ (I did not see another obvious location at the time). pkg/profiles/ would then subsample gridded fields from pkg/ecco/ like it already did for gridded fields from model/inc/ and now also does for pkg/ptracers/. In contrast, having pkg/ecco inherit gridded fields computed in pkg/profiles/ would have made no sense to me. Merging ecco/ and profiles/ would have been one way to avoid code duplication but gridded and ungridded data sets often involve different issues, methods, etc. so I am still not in favor of that approach. Another approach would consist in moving m_UE, m_VN, m_eta, etc. (computations and arrays) to model/ whenever they start being used in several packages -- if making profiles/ independent from ecco/ is important (?) then maybe this is the best way to proceed. Would be curious to know what @jm-c thinks about this.

@christophernhill, regarding "ecco, autodiff, prof etc.. are treading on each others toes": the situation may be worse than you realize. ctrl/ and various cryosphere related packages have their own gridded cost function codes and arrays -- maybe we should leave that alone for now, but all of this duplication would be worth looking into as some point. Things also remain pretty messy wrt ocbs/ controls and costs if I recall correctly -- let's leave that can of worms alone too. Not totally sure why you listed autodiff/ here but there probably was a good reason and cost/ probably should also be mentioned. If I were to summarize I would say that things have come a long way since the good old days when pkg/ecco/ contained a copy of the_main_loop, and I did put quite a bit of time removing the duplicates and disentangling some of this stuff, but a fair amount of work remains needed. @jrscott would do a great service to this community if he were to take on some of the remaining issues. I can provide some guidance, but again I would be curious to know what @jm-c would advise.

Sorry the text above got so long -- hopefully I spent enough time on it to avoid adding confusion...

christophernhill commented:
Hmmm - it sounds like there isn’t an easy way to do this more tidily as
things currently stands. Maybe logging the issues for things that might be
needed to make profs usable outside of Ecco-autodiff-curl-cost (assuming
that is desirable) is best that can be done for now and then doing ifndef
thing for now?

 It sounds like any approach except the ifndef Ecco scheme is going to
involve some effort.

On Tue, Mar 27, 2018 at 13:58 Gael Forget <notifications@github.com> wrote:

> @antnguyen13 <https://github.com/antnguyen13>: since this thread has
> grown so extensive already, I feel it may be worth asking: why is not
> compiling pkg/ecco/ important to you? Would this save you a large memory
> footprint even if data.ecco were empty? Something else?
>
> @christophernhill <https://github.com/christophernhill>, regarding "would
> be nice to avoid having a gradual accretion of duplicate stuff": yes that
> is my basic concern; replicating the m_eta computation and array in
> pkg/profiles/ would likely be next; and so on and so forth.
>
> @christophernhill <https://github.com/christophernhill>, regarding "ecco,
> autodiff, prof etc.. are treading on each others toes": I feel that ecco
> dealing with gridded data sets and profiles dealing with ungridded data
> sets makes sense as a general principle. The addition of ecco_phys.F was
> meant to provide a location for quantities like m_eta, m_UE, m_VN to be
> computed and stored within ecco/ (I did not see another obvious location at
> the time). pkg/profiles/ would then subsample gridded fields from pkg/ecco/
> like it already did for gridded fields from model/inc/ and now also does
> for pkg/ptracers/. In contrast, having pkg/ecco inherit gridded fields
> computed in pkg/profiles/ would have made no sense to me. Merging ecco/ and
> profiles/ would have been one way to avoid code duplication but gridded and
> ungridded data sets often involve different issues, methods, etc. so I am
> still not in favor of that approach. Another approach would consist in
> moving m_UE, m_VN, m_eta, etc. (computations and arrays) to model/ whenever
> they start being used in several packages -- if making profiles/
> independent from ecco/ is important (?) then maybe this is the best way to
> proceed. Would be curious to know what @jm-c <https://github.com/jm-c>
> thinks about this.
>
> @christophernhill <https://github.com/christophernhill>, regarding "ecco,
> autodiff, prof etc.. are treading on each others toes": the situation may
> be worse than you realize. ctrl/ and various cryosphere related packages
> have their own gridded cost function codes and arrays -- maybe we should
> leave that alone for now, but all of this duplication would be worth
> looking into as some point. Things also remain pretty messy wrt ocbs/
> controls and costs if I recall correctly -- let's leave that can of worms
> alone too. Not totally sure why you listed autodiff/ here but there
> probably was a good reason and cost/ probably should also be mentioned. If
> I were to summarize I would say that things have come a long way since the
> good old days when pkg/ecco/ contained a copy of the_main_loop, and I did
> put quite a bit of time removing the duplicates and disentangling some of
> this stuff, but a fair amount of work remains needed. @jrscott
> <https://github.com/jrscott> would do a great service to this community
> if he were to take on some of the remaining issues. I can provide some
> guidance, but again I would be curious to know what @jm-c
> <https://github.com/jm-c> would advise.
>
> Sorry the text above got so long -- hopefully I spent enough time on it to
> avoid adding confusion...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MITgcm/MITgcm/pull/71#issuecomment-376617935>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADXx4M3LEHHmIGaVdS7lVo4JdagbX7Mfks5tin2ngaJpZM4S5l7L>
> .
>

jm-c commented:
I took a look at pkg/profiles and changes from this PR:

1) appart from this '''profiles_inloop.F_new''', the rest seems reasonable;
 I also understand that one may want to use pkg/profiles without pkg/ecco,
 but I have no idea about what this means in term of cost/ctrl (and adjoint).
 @antnguyen13 does this work nicely in both forward and adjoint mode ?

2) for a coding point of view: 
 if the array to be sampled was passed as argument of S/R PROFILES_INTERP (instead
 of having all arrays included within profiles_interp.F) it would certainly
 help for this type of modification. But currently it's not the case
 so this modification has to involve saving m_UE and m_VN in a common block
 with the same name as in '''ecco.h'''.

On Tue, Mar 27, 2018 at 11:28:32AM -0700, christophernhill wrote:
> Hmmm - it sounds like there isn???t an easy way to do this more tidily as
> things currently stands. Maybe logging the issues for things that might be
> needed to make profs usable outside of Ecco-autodiff-curl-cost (assuming
> that is desirable) is best that can be done for now and then doing ifndef
> thing for now?
> 
>  It sounds like any approach except the ifndef Ecco scheme is going to
> involve some effort.
> 
> On Tue, Mar 27, 2018 at 13:58 Gael Forget <notifications@github.com> wrote:
> 
> > @antnguyen13 <https://github.com/antnguyen13>: since this thread has
> > grown so extensive already, I feel it may be worth asking: why is not
> > compiling pkg/ecco/ important to you? Would this save you a large memory
> > footprint even if data.ecco were empty? Something else?
> >
> > @christophernhill <https://github.com/christophernhill>, regarding "would
> > be nice to avoid having a gradual accretion of duplicate stuff": yes that
> > is my basic concern; replicating the m_eta computation and array in
> > pkg/profiles/ would likely be next; and so on and so forth.
> >
> > @christophernhill <https://github.com/christophernhill>, regarding "ecco,
> > autodiff, prof etc.. are treading on each others toes": I feel that ecco
> > dealing with gridded data sets and profiles dealing with ungridded data
> > sets makes sense as a general principle. The addition of ecco_phys.F was
> > meant to provide a location for quantities like m_eta, m_UE, m_VN to be
> > computed and stored within ecco/ (I did not see another obvious location at
> > the time). pkg/profiles/ would then subsample gridded fields from pkg/ecco/
> > like it already did for gridded fields from model/inc/ and now also does
> > for pkg/ptracers/. In contrast, having pkg/ecco inherit gridded fields
> > computed in pkg/profiles/ would have made no sense to me. Merging ecco/ and
> > profiles/ would have been one way to avoid code duplication but gridded and
> > ungridded data sets often involve different issues, methods, etc. so I am
> > still not in favor of that approach. Another approach would consist in
> > moving m_UE, m_VN, m_eta, etc. (computations and arrays) to model/ whenever
> > they start being used in several packages -- if making profiles/
> > independent from ecco/ is important (?) then maybe this is the best way to
> > proceed. Would be curious to know what @jm-c <https://github.com/jm-c>
> > thinks about this.
> >
> > @christophernhill <https://github.com/christophernhill>, regarding "ecco,
> > autodiff, prof etc.. are treading on each others toes": the situation may
> > be worse than you realize. ctrl/ and various cryosphere related packages
> > have their own gridded cost function codes and arrays -- maybe we should
> > leave that alone for now, but all of this duplication would be worth
> > looking into as some point. Things also remain pretty messy wrt ocbs/
> > controls and costs if I recall correctly -- let's leave that can of worms
> > alone too. Not totally sure why you listed autodiff/ here but there
> > probably was a good reason and cost/ probably should also be mentioned. If
> > I were to summarize I would say that things have come a long way since the
> > good old days when pkg/ecco/ contained a copy of the_main_loop, and I did
> > put quite a bit of time removing the duplicates and disentangling some of
> > this stuff, but a fair amount of work remains needed. @jrscott
> > <https://github.com/jrscott> would do a great service to this community
> > if he were to take on some of the remaining issues. I can provide some
> > guidance, but again I would be curious to know what @jm-c
> > <https://github.com/jm-c> would advise.
> >
> > Sorry the text above got so long -- hopefully I spent enough time on it to
> > avoid adding confusion...
> >
> > ???
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/MITgcm/MITgcm/pull/71#issuecomment-376617935>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/ADXx4M3LEHHmIGaVdS7lVo4JdagbX7Mfks5tin2ngaJpZM4S5l7L>
> > .
> >
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/71#issuecomment-376627809

christophernhill commented:
@antnguyen13 I did
```
git clone https://github.com/antnguyen13/MITgcm antnguyen13-MITgcm
cd antnguyen13-MITgcm
git checkout separate_profiles_ecco
cd pkg/profiles/
git rm profiles_inloop.F_new
git commit
git push
```
to get rid of zoombie ```profiles_inloop.F_new``` in PR (it will still hang around in intermediate commit but thats OK). 
The file may still be lurking in other branches you have and/or in local copies that haven't
been updated. I think that will do for now - lets hope it doesn't come back from the dead....

gaelforget commented:
In fact, pkg/profiles/ has long been tested without any of the state estimation packages in global_with_exf (in forward mode when pkg/profiles/ can be used as a diagnostic tool) and it seems to behave as expected in that configuration. @antnguyen13's issue rather is that variables which live in pkg/ecco/ are not available to pkg/profiles/ when pkg/ecco/ is not compiled (my guess is that this currently results in a bunch of 0s being outputed without any error or warning). 

With this being said, I just double checked: pkg/ecco/ does have a sizable memory footprint based on what the ECCO_OPTIONS.h and ecco.h defaults. So here is what I would propose: if agreed that, once the memory footprint of pkg/ecco/ has been reduced, the newly duplicated code will be removed and replaced with a warning that pkg/ecco should be compiled (or a stop) then just proceed with the PR and let's be done with this for now.

christophernhill commented:
Made some notes in issue #79 and merging.
